File: app/src/main/java/com/wyang/study/ui/fragment_second/StickyDecorationFragment.java
package com.wyang.study.ui.fragment_second;

import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Rect;
import android.graphics.drawable.GradientDrawable;
import android.support.annotation.NonNull;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.text.TextUtils;
import android.view.View;
import android.widget.TextView;

import com.chad.library.adapter.base.BaseQuickAdapter;
import com.chad.library.adapter.base.BaseViewHolder;
import com.gavin.com.library.PowerfulStickyDecoration;
import com.gavin.com.library.listener.PowerGroupListener;
import com.wyang.study.R;
import com.wyang.study.ui.base.BaseFragment;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;

/**
 * Created by weiyang on 2019-10-16.
 */
public class StickyDecorationFragment extends BaseFragment {
    @BindView(R.id.mRecyclerView)
    RecyclerView mRecyclerView;
    private StickyAdapter mAdapter;

    @Override
    protected int getLayoutId() {
        return R.layout.fragment_sticky_decoration;
    }

    @Override
    protected void initView() {
        mRecyclerView.setLayoutManager(new LinearLayoutManager(getContext()));

        PowerGroupListener listener = new PowerGroupListener() {
            @Override
            public String getGroupName(int position) {
                String item = mAdapter.getData().get(position);
                return item.substring(0, 3);
            }

            @Override
            public View getGroupView(int position) {
                //获取自定定义的组View
                View view = getLayoutInflater().inflate(R.layout.sticky_header, null, false);
                TextView tv_type = view.findViewById(R.id.tv_type);
                tv_type.setText(getGroupName(position));

                return view;
            }
        };

        View v = getLayoutInflater().inflate(R.layout.sticky_header, null, false);
        measureWidthAndHeight(v);

        PowerfulStickyDecoration decoration = PowerfulStickyDecoration.Builder
                .init(listener)
                .setDivideHeight(0)
                .setDivideColor(0xffe2e2e2)
                .setGroupHeight(dp2px(40))
                .setGroupBackground(0x00000000)
                //重置span（注意：使用GridLayoutManager时必须调用）
                //.resetSpan(mRecyclerView, (GridLayoutManager) manager)
                .build();

        mRecyclerView.addItemDecoration(new ItemDecoration());
        mRecyclerView.addItemDecoration(decoration);

        mAdapter = new StickyAdapter();
        mAdapter.bindToRecyclerView(mRecyclerView);

        List<String> list = new ArrayList<>();
        list.add("分组1-AAAAA");
        list.add("分组1-BBBBB");
        list.add("分组1-CCCCC");

        list.add("分组2-AAAAA");
        list.add("分组2-BBBBB");
        list.add("分组2-CCCCC");

        list.add("分组3-AAAAA");
        list.add("分组3-AAAAA");
        list.add("分组3-AAAAA");

        list.add("分组4-DDDDD");
        list.add("分组4-EEEEE");
        list.add("分组4-EEEEE");

        list.add("分组5-FFFFF");
        list.add("分组5-FFFFF");
        list.add("分组5-FFFFF");
        list.add("分组5-GGGGG");
        mAdapter.setNewData(list);
    }

    public void measureWidthAndHeight(View view) {
        int w = View.MeasureSpec.makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED);
        int h = View.MeasureSpec.makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED);
        view.measure(w, h);
    }

    /**
     * 判断是不是分组第一个
     */
    private boolean isFirst(int pos) {
        if (pos == 0)
            return true;

        String c = mAdapter.getData().get(pos);
        String p = mAdapter.getData().get(pos - 1);

        boolean rst = !TextUtils.equals(c.substring(0, 3), p.substring(0, 3));
        return rst;
    }

    /**
     * 判断是不是分组最后一个
     */
    private boolean isFooter(int pos) {
        int size = mAdapter.getData().size();
        if (pos == size - 1)
            return true;
        String c = mAdapter.getData().get(pos);
        String n = mAdapter.getData().get(pos + 1);
        return !TextUtils.equals(c.substring(0, 3), n.substring(0, 3));
    }

    class StickyAdapter extends BaseQuickAdapter<String, BaseViewHolder> {

        public StickyAdapter() {
            super(R.layout.item_sticky_recycler);
        }

        @Override
        protected void convert(BaseViewHolder helper, String item) {
            helper.setText(R.id.tv_content, item);
        }
    }

    public int dp2px(float dp) {
        final float scale = getResources().getDisplayMetrics().density;
        return (int) (dp * scale + 0.5f);
    }

    class ItemDecoration extends RecyclerView.ItemDecoration {
        private GradientDrawable topCorner;
        private GradientDrawable bottomCorner;

        public ItemDecoration() {
            int radius = dp2px(10);

            topCorner = new GradientDrawable();
            topCorner.setShape(GradientDrawable.RECTANGLE);
            topCorner.setCornerRadii(new float[]{
                    radius, radius,
                    radius, radius,
                    0, 0,
                    0, 0
            });
            topCorner.setColor(Color.WHITE);

            bottomCorner = new GradientDrawable();
            bottomCorner.setShape(GradientDrawable.RECTANGLE);
            bottomCorner.setCornerRadii(new float[]{
                    0, 0,
                    0, 0,
                    radius, radius,
                    radius, radius
            });
            bottomCorner.setColor(Color.WHITE);

        }

        @Override
        public void getItemOffsets(@NonNull Rect outRect, @NonNull View view, @NonNull RecyclerView parent, @NonNull RecyclerView.State state) {
            int position = parent.getChildAdapterPosition(view);
            if (position == 0) {
                outRect.top = dp2px(20);
            } else if (isFirst(position)) {
                outRect.top = dp2px(30);
            } else {
                outRect.top = dp2px(0);
            }

            if (position == mAdapter.getData().size() - 1) {
                outRect.bottom = dp2px(20);
            } else {
                outRect.bottom = 0;
            }
        }

        @Override
        public void onDraw(@NonNull Canvas canvas, @NonNull RecyclerView parent, @NonNull RecyclerView.State state) {
            super.onDraw(canvas, parent, state);

            int childCount = parent.getChildCount();
            for (int i = 0; i < childCount; i++) {
                View childAt = parent.getChildAt(i);
                int position = parent.getChildLayoutPosition(childAt);
                int left = parent.getPaddingLeft();
                int right = parent.getWidth() - parent.getPaddingRight();
                int top = childAt.getTop() - dp2px(40);
                if (position == 0 || position == mAdapter.getData().size() - 1) {
                    if (position == 0) {
                        top -= dp2px(20);
                        topCorner.setBounds(left, top + dp2px(10), right, top + dp2px(20));
                        topCorner.draw(canvas);
                    } else {
                        top = childAt.getBottom();
                        bottomCorner.setBounds(left, top, right, top + dp2px(10));
                        bottomCorner.draw(canvas);
                    }
                } else {
                    top -= dp2px(30);

                    if (isFirst(position)) {   // 是分组第一个，则绘制
                        bottomCorner.setBounds(left, top, right, top + dp2px(10));
                        bottomCorner.draw(canvas);

                        topCorner.setBounds(left, top + dp2px(20), right, top + dp2px(30));
                        topCorner.draw(canvas);
                    }
                }
            }
        }
    }
}

File: app/src/main/java/com/wyang/study/ui/fragment_second/StickyDecorationFragment.kt
package com.wyang.study.ui.fragment_second

import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Rect
import android.graphics.drawable.GradientDrawable
import android.text.TextUtils
import android.view.View
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.chad.library.adapter.base.BaseQuickAdapter
import com.chad.library.adapter.base.BaseViewHolder
import com.gavin.com.library.PowerfulStickyDecoration
import com.gavin.com.library.listener.PowerGroupListener
import com.wyang.study.R
import com.wyang.study.databinding.FragmentStickyDecorationBinding
import com.wyang.study.databinding.WidgetRecyclerBinding
import com.wyang.study.ui.base.BaseFragment

class StickyDecorationFragment : BaseFragment<FragmentStickyDecorationBinding>() {
    private var mRecyclerBinding: WidgetRecyclerBinding? = null
    private var mAdapter: StickyAdapter? = null

    override fun getViewBinding(): FragmentStickyDecorationBinding {
        return FragmentStickyDecorationBinding.inflate(layoutInflater)
    }

    override fun initialize() {
        mBinding?.root?.let { mRecyclerBinding = WidgetRecyclerBinding.bind(it) }

        val listener: PowerGroupListener = object : PowerGroupListener {
            override fun getGroupName(position: Int): String? {
                val item: String? = mAdapter?.data?.get(position)
                return item?.substring(0, 3)
            }

            override fun getGroupView(position: Int): View {
                //获取自定定义的组View
                val view: View = layoutInflater.inflate(R.layout.sticky_header, null, false)
                val tv_type = view.findViewById<TextView>(R.id.tv_type)
                tv_type.text = getGroupName(position)
                return view
            }
        }

        val v: View = layoutInflater.inflate(R.layout.sticky_header, null, false)
        measureWidthAndHeight(v)

        val decoration = PowerfulStickyDecoration.Builder
            .init(listener)
            .setDivideHeight(0)
            .setDivideColor(-0x1d1d1e)
            .setGroupHeight(dp2px(40f)) //分组的高度
            .setGroupBackground(0x00000000) //重置span（注意：使用GridLayoutManager时必须调用）
            //.resetSpan(mRecyclerView, (GridLayoutManager) manager)
            .build()

        mRecyclerBinding?.mRecyclerView?.addItemDecoration(ItemDecoration())
        mRecyclerBinding?.mRecyclerView?.addItemDecoration(decoration)

        mAdapter = StickyAdapter()
        mAdapter?.bindToRecyclerView(mRecyclerBinding?.mRecyclerView)

        val list: MutableList<String> = ArrayList()
        list.add("分组1-AAAAA")
        list.add("分组1-BBBBB")
        list.add("分组1-CCCCC")

        list.add("分组2-AAAAA")
        list.add("分组2-BBBBB")
        list.add("分组2-CCCCC")

        list.add("分组3-AAAAA")
        list.add("分组3-AAAAA")
        list.add("分组3-AAAAA")

        list.add("分组4-DDDDD")
        list.add("分组4-EEEEE")
        list.add("分组4-EEEEE")

        list.add("分组5-FFFFF")
        list.add("分组5-FFFFF")
        list.add("分组5-FFFFF")
        list.add("分组5-GGGGG")
        mAdapter?.setNewData(list as List<String?>?)
    }

    fun measureWidthAndHeight(view: View) {
        val w = View.MeasureSpec.makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED)
        val h = View.MeasureSpec.makeMeasureSpec(0, View.MeasureSpec.UNSPECIFIED)
        view.measure(w, h)
    }

    /**
     * 判断是不是分组第一个
     */
    private fun isFirst(pos: Int): Boolean {
        if (pos == 0) return true
        val c: String? = mAdapter?.data?.get(pos)
        val p: String? = mAdapter?.data?.get(pos - 1)
        return !TextUtils.equals(c?.substring(0, 3), p?.substring(0, 3))
    }

    /**
     * 判断是不是分组最后一个
     */
    private fun isFooter(pos: Int): Boolean {
        val size: Int? = mAdapter?.data?.size
        if (size != null) {
            if (pos == size - 1) return true
        }
        val c: String? = mAdapter?.data?.get(pos)
        val n: String? = mAdapter?.data?.get(pos + 1)
        if (c != null) {
            return !TextUtils.equals(c.substring(0, 3), n?.substring(0, 3))
        }
        return false
    }

    class StickyAdapter : BaseQuickAdapter<String?, BaseViewHolder>(R.layout.item_sticky_recycler) {
        override fun convert(helper: BaseViewHolder, item: String?) {
            helper.setText(R.id.tv_content, item)
        }
    }

    fun dp2px(dp: Float): Int {
        val scale = resources.displayMetrics.density
        return (dp * scale + 0.5f).toInt()
    }

    inner class ItemDecoration : RecyclerView.ItemDecoration() {
        private val topCorner: GradientDrawable
        private val bottomCorner: GradientDrawable
        override fun getItemOffsets(
            outRect: Rect,
            view: View,
            parent: RecyclerView,
            state: RecyclerView.State
        ) {
            val position = parent.getChildAdapterPosition(view)
            when {
                position == 0 -> {
                    outRect.top = dp2px(20f)
                }
                isFirst(position) -> {
                    outRect.top = dp2px(30f)
                }
                else -> {
                    outRect.top = dp2px(0f)
                }
            }
            if (position == mAdapter?.data?.size?.minus(1) ?: 0) {
                outRect.bottom = dp2px(20f)
            } else {
                outRect.bottom = 0
            }
        }

        override fun onDraw(canvas: Canvas, parent: RecyclerView, state: RecyclerView.State) {
            super.onDraw(canvas, parent, state)
            val childCount = parent.childCount
            for (i in 0 until childCount) {
                val childAt = parent.getChildAt(i)
                val position = parent.getChildLayoutPosition(childAt)
                val left = parent.paddingLeft
                val right = parent.width - parent.paddingRight
                var top: Int = childAt.top - dp2px(40f) //减去分组的高度
                if (position == 0 || position == mAdapter?.data?.size?.minus(1) ?: 0) {
                    if (position == 0) {
                        top -= dp2px(20f)
                        topCorner.setBounds(left, top + dp2px(10f), right, top + dp2px(20f))
                        topCorner.draw(canvas)
                    } else {
                        top = childAt.bottom
                        bottomCorner.setBounds(left, top, right, top + dp2px(10f))
                        bottomCorner.draw(canvas)
                    }
                } else {
                    top -= dp2px(30f)
                    if (isFirst(position)) {   // 是分组第一个，则绘制
                        bottomCorner.setBounds(left, top, right, top + dp2px(10f))
                        bottomCorner.draw(canvas)
                        topCorner.setBounds(left, top + dp2px(20f), right, top + dp2px(30f))
                        topCorner.draw(canvas)
                    }
                }
            }
        }

        init {
            val radius: Int = dp2px(10f)
            topCorner = GradientDrawable()
            topCorner.shape = GradientDrawable.RECTANGLE
            topCorner.cornerRadii = floatArrayOf(
                radius.toFloat(), radius.toFloat(),
                radius.toFloat(), radius.toFloat(), 0f, 0f, 0f, 0f
            )
            topCorner.setColor(Color.WHITE)
            bottomCorner = GradientDrawable()
            bottomCorner.shape = GradientDrawable.RECTANGLE
            bottomCorner.cornerRadii = floatArrayOf(
                0f, 0f, 0f, 0f,
                radius.toFloat(), radius.toFloat(),
                radius.toFloat(), radius
                    .toFloat()
            )
            bottomCorner.setColor(Color.WHITE)
        }
    }
}